Migrate backends/mediatek to the new namespace

dbort opened the pull request:
Summary: Move the mediatek backend out of the torch:: namespace, and update to avoid using the torch:: or exec_aten:: namespaces.

Differential Revision: D63928380

dbort commented:
Updated this PR to avoid changing the namespace of headers that do not have .cpp files in the executorch repo, because they're defined in the prebuilt library.

I successfully built and linked the version of the code in this PR using the instructions in the README.

dbort commented on backends/mediatek/runtime/include/NeuronBufferAllocator.h:
```diff
@@ -102,7 +107,7 @@ class MemoryUnit {
 };
 
-class BufferAllocator : public MemoryAllocator {
+class BufferAllocator : public executorch::runtime::MemoryAllocator {
```
It looks like there is prebuilt code that refers to ET (executorch) types: specifically, this subclass of MemoryAllocator.

ET does not guarantee ABI compatibility across versions, or even between git commits. If the definition of the MemoryAllocator base class were to change in an incompatible way (new field, different vtable order), then BufferAllocator would misbehave when interpreted as a `MemoryAllocator*`: at best, crashes; at worst, subtle memory corruption or execution of random code.

It is not safe for prebuilt code to refer to ET types.

neuropilot-captain replied:
Thank you for bringing this to our attention. We are currently conducting an internal review of the library's source code, which will be made publicly available upon completion. This will eliminate the need for prebuilt binaries.
In the meantime, we will establish CI to ensure MTK backend works with the OSS models.

dbort replied:
I created https://github.com/pytorch/executorch/issues/6074 to track this issue. Please comment on that task if you have any questions for me!

cccclai reviewed (approved):
Looks good and @dbort has tested briefly. @neuropilot-captain let us know if we miss something

neuropilot-captain reviewed (approved):
Looks good!

dbort committed:
Migrate backends/mediatek to the new namespace (#5956)

Summary:

Move the mediatek backend out of the torch:: namespace, and update to avoid using the torch:: or exec_aten:: namespaces.

Reviewed By: cccclai

Differential Revision: D63928380

dbort commented:
@pytorchbot cherry-pick --onto release/0.4 -c fixnewfeature
